File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/wheel/GroundReactionAccelerationCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.wheel;

import com.github.aadvorak.artilleryonline.battle.calculations.WheelCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.WheelGroundState;
import com.github.aadvorak.artilleryonline.battle.common.VectorProjections;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;

public class GroundReactionAccelerationCalculator {

    public static void calculate(WheelCalculations wheelCalculations, VehicleModel vehicleModel,
                                 double groundReactionCoefficient) {
        if (WheelGroundState.FULL_OVER_GROUND.equals(wheelCalculations.getGroundState())
                || WheelGroundState.FULL_UNDER_GROUND.equals(wheelCalculations.getGroundState())) {
            return;
        }
        var groundAngle = wheelCalculations.getGroundAngle();
        var velocityNormalProjection = wheelCalculations.getVelocity().projections(groundAngle).getNormal();
        if (velocityNormalProjection < 0) {
            var mass = vehicleModel.getPreCalc().getMass();
            var accelerationProjections = new VectorProjections(groundAngle)
                    .setNormal(- velocityNormalProjection * wheelCalculations.getGroundDepth()
                            * groundReactionCoefficient / mass);
            wheelCalculations.setGroundReactionAcceleration(accelerationProjections.recoverAcceleration());
        }
    }
}

File: frontend/playground/battle/calculator/wheel/ground-reaction-acceleration-calculator.ts
import {type WheelCalculations, WheelGroundState} from '@/playground/data/calculations'
import {VectorUtils} from '@/playground/utils/vector-utils'
import type {VehicleModel} from "~/playground/data/model";

export const GroundReactionAccelerationCalculator = {
  calculate(wheelCalculations: WheelCalculations, vehicleModel: VehicleModel, groundReactionCoefficient: number): void {
    if (
        WheelGroundState.FULL_OVER_GROUND === wheelCalculations.groundState ||
        WheelGroundState.FULL_UNDER_GROUND === wheelCalculations.groundState
    ) {
      return
    }

    const groundAngle = wheelCalculations.groundAngle!
    const velocityVerticalProjection = VectorUtils.getVerticalProjection(
        wheelCalculations.velocity!,
        groundAngle
    )

    if (velocityVerticalProjection < 0) {
      const mass = vehicleModel.preCalc.mass
      const accelerationVerticalProjection =
          -velocityVerticalProjection * wheelCalculations.groundDepth! * groundReactionCoefficient / mass

      wheelCalculations.groundReactionAcceleration.x = VectorUtils.getComponentX(
          accelerationVerticalProjection,
          0.0,
          groundAngle
      )
      wheelCalculations.groundReactionAcceleration.y = VectorUtils.getComponentY(
          accelerationVerticalProjection,
          0.0,
          groundAngle
      )
    }
  }
}
